### PR TITLE
feat: Remove GameStateCell's T: Clone requirement

### PR DESCRIPTION
**Pull request by caspark**
It turns out that the restriction of only storing cloneable things in `GameStateCell` is not actually necessary, since it's trivial for client code to do the actual cloning of the data stored in `GameStateCell` if they have access to it.
    
For convenience, to avoid breaking API, and to steer users away from the footgun of modifying a stored state, we still provide a `load()` function for the case where `T` is `Clone`.

---

This was originally motivated by me needing to clone a type which is faster to clone if I have a mutable reference to it; I could have used interior mutability via `UnsafeCell` but just out of interest I tried seeing if ggrs's `Clone` bound for game state could be relaxed, and it turns out that it can.

It seemed like a change that might be worth upstreaming since it simplies the code a little, and since bounds on impls are [often more ergonomic](https://stackoverflow.com/a/66369912/775982) than bounds on structs - and perhaps there are other niche use cases for being able to read data in a GameStateCell without cloning it that this enables too.

Note that this does have one downside: `MappedMutexGuard` becomes part of the public API, whereas before it was only an implementation detail. If that's a problem then the guard could be newtyped in a custom type with its own Deref(Mut) implementation, but I figure exposing the MappedMutexGuard directly is more flexible.

### CODE REVIEW

**Comment by gschup** [comment]
Thanks for the PR! I like the change, even if it explicitly shows `MappedMutexGuard`. Since the examples hold up and the old way still exists, this should give experienced users more power!

File: src/sync_layer.rs
Comment (gschup) [inline src/sync_layer.rs:32]: Is there an option to prevent the misuse of this? Any way of giving a non-mutable reference?

Comment (caspark) [reply src/sync_layer.rs]: I have pushed [eb1bebd](https://github.com/gschup/ggrs/pull/80/commits/eb1bebd2ed60beb46bfb37e26c5d44b8e06b5029) which implements the newtype pattern I mentioned in the PR's description (see commit description).

With that in place, the default is a shared reference, rather than a mutable reference - so possible misuse is not possible by accident (it has to be very explicitly opted into). It's a little more code but I like the result better than the original PR I raised; feel free to take it or leave it.

If you meant "is there an alternate parking_lot API we could use instead": I don't think it's possible to do this restriction without a newtype in our code, as there isn't any way to "unimplement" `DerefMut` for `MappedMutexGuard`, and there's no way I could see to downgrade a `MappedMutexGuard` to a hypothetical `SharedMutexGuard` either.

**Comment by gschup** [comment]
I like  `.as_mut_dangerous()`, it is quite clear on what can happen if you mess up while using it :)

**Review by gschup — APPROVED** [review]
I think we can merge this for now and see if problems arise along the way. Great work!

**Pull request merged**
